Only call render function when window is visible for macOS and X11

trimental opened the pull request:
This functionality depends upon a PR I've put in for winit, but I'm sending this PR now to check if it's viable for this crate.

Basically it makes it so that on macOS the render function is only called when the winit window is visible, whilst keeping the update function the same. The reasons for this are A) the game loop can sleep in-between calls to update which significantly saves cpu time, B) since the rendering code won't be called, it significantly saves either cpu and/or gpu time; provided the window is not visible to the user.

In a non scientific test, my games go from 12-18% cpu usage, when the window is in the background, to about 3-4%.

trimental committed:
Only call render function when window is visible

tuzz commented:
Hey, sorry, I'll try to review this soon. It looks good and I think I've experienced this problem myself.

trimental commented:
@tuzz No worries. Just updated the PR to update to winit, fix wasm as it doesn't support sleeping, and removed checking for occlusion when deciding if to sleep before calling update.

trimental committed:
Fix wasm and update winit to 0.27

tuzz commented:
Hi @trimental, sorry that's it taken me so long to reply.

I'm keen to merge this change in, I'm just wondering if this implementation would be simpler?

```rust
if g.occluded && cfg!(not(target_arch = "wasm32")) {
  sleep(Duration::from_secs_f64(g.fixed_time_step))
} else {
  render(g);
}
```

My thinking is that we sleep for the fixed_time_step if the window is occluded, rather than rendering. On WASM, we call render as normal since the browser should remain in control of deciding when to re-render the canvas.

I thought about making the behavior when occluded configuarable, but I don't think there's a strong enough use case for continuing to render when the window is occluded, so I think it's simpler to just make this the standard behavior.

What do you think? Thanks

tuzz commented:
I released the slightly simpler approach in version `1.0.0` (see [commit](https://github.com/tuzz/game-loop/commit/58c21bddfaaf007831b96887f3dd2746927eeb8f)). Please re-open this PR if you think we should change the implementation. Thanks for your contribution!
